query_executor.go

```diff
@@ -115,45 +115,70 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
 
 	var lastErr error
 	var iter *Iter
+	var conn *Conn
+	var potentiallyExecuted bool
 	for selectedHost != nil {
 		host := selectedHost.Info()
 		if host == nil || !host.IsUp() {
-			selectedHost = hostIter()
-			continue
-		}
-
-		pool, ok := q.pool.getPool(host)
-		if !ok {
-			selectedHost = hostIter()
-			continue
-		}
-
-		conn := pool.Pick(selectedHost.Token(), qry)
-		if conn == nil {
-			selectedHost = hostIter()
-			continue
-		}
-
-		iter = q.attemptQuery(ctx, qry, conn)
-		iter.host = selectedHost.Info()
-		// Update host
-		switch iter.err {
-		case context.Canceled, context.DeadlineExceeded, ErrNotFound:
-			// those errors represents logical errors, they should not count
-			// toward removing a node from the pool
-			selectedHost.Mark(nil)
-			return iter
-		default:
-			selectedHost.Mark(iter.err)
-		}
-
-		// Exit if the query was successful
-		// or no retry policy defined
-		if iter.err == nil || rt == nil {
-			return iter
+			if rt == nil {
+				selectedHost = hostIter()
+				lastErr = ErrHostDown
+				continue
+			}
+		} else {
+			pool, ok := q.pool.getPool(host)
+			if !ok {
+				if rt == nil {
+					selectedHost = hostIter()
+					lastErr = ErrNoPool
+					continue
+				}
+			} else {
+				conn = pool.Pick(selectedHost.Token(), qry)
+				if conn == nil {
+					if rt == nil {
+						selectedHost = hostIter()
+						lastErr = ErrNoConnectionsInPool
+						continue
+					}
+				} else {
+					iter = q.attemptQuery(ctx, qry, conn)
+					iter.host = selectedHost.Info()
+					// Update host
+					switch iter.err {
```

Comment:
It won't work, we wrap it in `QueryError`, now we need to do `errors.Is` instead